File: src/main/java/dsa/PreditWinner.java
package dsa;

public class PreditWinner {
    public static  boolean predictTheWinner(int[] nums) {
        int n = nums.length;
        // Use Integer to allow null (indicating "not computed")
        Integer[][] dp = new Integer[n][n];
        // Start the recursion from the full array range
        return maxDiff(nums, 0, n - 1, dp) >= 0;
    }

    private static int maxDiff(int[] nums, int left, int right, Integer[][] dp) {
        // Base case: only one element left
        if (left == right) {
            return nums[left];
        }
        // Return memoized result if already computed
        if (dp[left][right] != null) {
            return dp[left][right];
        }
        // Choose the left element and compute the difference
        int leftChoice = nums[left] - maxDiff(nums, left + 1, right, dp);
        // Choose the right element and compute the difference
        int rightChoice = nums[right] - maxDiff(nums, left, right - 1, dp);
        // Store and return the maximum difference
        dp[left][right] = Math.max(leftChoice, rightChoice);
        return dp[left][right];
    }

    public  static  void  main(String [] ar) {
        System.out.println(predictTheWinner(new int[]{1,5,2}));
    }
}

File: src/jsImpl/preditWinner.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
// var predictTheWinner = function(nums) {
//     var p1Sum = 0 
//     var p2Sum = 0
//     function pick(num,which,type){
//         if(num.length == 1){
//             if(which == "0"){
//                 p1Sum = p1Sum + (type == "start" ? num.shift() : num.pop())
//             }
//             if(which == "1"){
//                 p2Sum = p2Sum +(type == "start" ? num.shift() : num.pop()  )
//             }
//             return
//         }
//         if(which == "0"){
//             p1Sum = p1Sum + (type == "start" ? num.shift() : num.pop())
//             pick(num,"1","end")
//         }
//         if(which == "1"){
//             p2Sum = p2Sum + (type == "start" ? num.shift() : num.pop()  )
//             pick(num,"0","start")
//         }
//     }

//     // function pickRight(numm,which){
//     //     if(numm.length == 1){
//     //         if(which == "0"){
//     //             p1Sum = p1Sum + numm.pop()
//     //         }
//     //         if(which == "1"){
//     //             p2Sum = p2Sum + numm.pop()  
//     //         }
//     //         return
//     //     }
//     //     if(which == "0"){
//     //         p1Sum = p1Sum + numm.pop()
//     //         pickRight(numm,"1")
//     //     }
//     //     if(which == "1"){
//     //         p2Sum = p2Sum + numm.pop()  
//     //         pickRight(numm,"0")
//     //     }
//     // }
//     pick([...nums],"0","start")
//     console.log(p1Sum,p2Sum)
//     if(p1Sum > p2Sum){
//         return true
//     }
//     p1Sum = 0
//     p2Sum = 0
//     pick([...nums],"0","end")
//     if(p1Sum > p2Sum){
//         return true
//     }
//     console.log(p1Sum,p2Sum)

//     return false
// };


// predictTheWinner([2,4,55,6,8])


var predictTheWinner = function(nums){
    const dp  = []
    for(let i =0 ;i<nums.length;i++){
        dp[i] = Array(nums.length).fill(-1)
    }
    function maxDiff(nums,left,right,n){
        if(left == right){
            return nums[left]
        }
        if(dp[left][right] != -1){
            return dp[left][right]
        }
        const sl = nums[left] - maxDiff(nums,left+1,right,nums.length)
        const sr = nums[right] - maxDiff(nums,left,right-1,nums.length)
        dp[left][right] = Math.max(sl,sr)
        return Math.max(sl,sr)
    }

    return maxDiff(nums,0,nums.length-1,nums.length) >= 0 ? true : false
 }

console.log( predictTheWinner([1,5,233,7]))